migrate.go

```diff
@@ -128,30 +128,86 @@ func (ms Migrations) String() string {
 // GoMigration is a Go migration func that is run within a transaction.
 type GoMigration func(tx *sql.Tx) error
 
+// GoMigrationContext is a Go migration func that is run within a transaction and receives a context.
+type GoMigrationContext func(ctx context.Context, tx *sql.Tx) error
+
 // GoMigrationNoTx is a Go migration func that is run outside a transaction.
 type GoMigrationNoTx func(db *sql.DB) error
 
+// GoMigrationNoTxContext is a Go migration func that is run outside a transaction and receives a context.
+type GoMigrationNoTxContext func(ctx context.Context, db *sql.DB) error
+
+// withoutContext wraps a GoMigration to make it a GoMigrationContext.
+func (gm GoMigrationContext) withoutContext() GoMigration {
+	return func(tx *sql.Tx) error {
+		return gm(context.Background(), tx)
+	}
+}
+
+// withContext wraps a GoMigration to make it a GoMigrationContext.
+func (gm GoMigrationNoTxContext) withoutContext() GoMigrationNoTx {
+	return func(db *sql.DB) error {
+		return gm(context.Background(), db)
+	}
+}
+
+// withContext wraps a GoMigration to make it a GoMigrationContext.
+func (gm GoMigration) withContext() GoMigrationContext {
+	return func(_ context.Context, tx *sql.Tx) error {
+		return gm(tx)
+	}
+}
+
+// withContext wraps a GoMigrationNoTx to make it a GoMigrationNoTxContext.
+func (gm GoMigrationNoTx) withContext() GoMigrationNoTxContext {
+	return func(_ context.Context, db *sql.DB) error {
+		return gm(db)
+	}
+}
+
 // AddMigration adds Go migrations.
 func AddMigration(up, down GoMigration) {
 	_, filename, _, _ := runtime.Caller(1)
-	AddNamedMigration(filename, up, down)
+	// intentionally don't call to AddMigrationContext so each of these functions can calculate the filename correctly
+	AddNamedMigrationContext(filename, up.withContext(), down.withContext())
```

Comment:
I'm not sure this is correct, because users can pass `nil` functions, wouldn't this panic?

Everything in this PR looks fine, it's just the Go-based migration code that seems a bit complicated (maybe it's just me, need to spend a bit more time making sure this works as expected)

Reply:
@ori-shalom Can you revert the Go changes and we can tackle them as a follow-up PR?

Here's the revert, but I'm sure you can handle it.

https://github.com/piiano/goose/commit/9bf933da0b99b0b8e35ef2f804fcf6a2dbaa4382

Reply:
Thanks for the review @mfridman.
I added a revert commit + rebased to resolve conflicts.
I'll open another PR to add context support to migrations.

Reply:
@mfridman I added #534 as a separate PR that adds context to migrations.
It also include a fix to the issue you raised about passing a `nil` function for up/down.